src/components/Form/Form.js

```diff
@@ -0,0 +1,78 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+import { useFormState } from '../../hooks/useFormState'
+
+export function Form({ children, config, render }) {
+  const inputNames = Object.keys(config.inputs)
+
+  // Set up initial values
+  let initialValues = {}
+  inputNames.forEach((x) => {
+    initialValues[x] = 'err' // by default inputs have "hidden" errors
+  })
+
+  // Hooks
+  const [
+    getInputErrors,
+    getInputValues,
+    setStateFactory,
+    getFormErrorMessage,
+    setFormErrorMessage,
+    getFormIsValid,
+  ] = useFormState(initialValues)
+
+  // Validates form, sets button to disabled if there's an error
+  function inputNamePropsTransformer(inputName) {
+    return {
+      name: inputName,
+      formChangeHandler: setStateFactory(inputName),
+      validator: (input) =>
+        config.inputs[inputName].validators
+          .reduce((errors, validator) => errors.concat(validator(input)), [])
+          .filter((x) => !!x) // remove empty strings
+          .join(', '),
+      labelCopy: config.inputs[inputName].labelCopy,
+      ['data-tid']: [config.formName, config.formId, inputName].join('-'),
+    }
+  }
+
+  async function onSubmit(syntheticReactEvent) {
+    setFormErrorMessage('')
+
+    syntheticReactEvent.preventDefault()
+
+    if (!getFormIsValid()) {
+      // Should already be handled by form itself, but...
+      if (getInputErrors()) {
+        setFormErrorMessage(
+          'One of the inputs was invalid. Errors: ' + getInputErrors()
+        )
+      } else {
+        setFormErrorMessage("You haven't typed anything yet")
+      }
+      return
+    }
+
+    try {
+      await config.onSubmit(getInputValues())
+      alert(
+        'form submission successful with values:' +
+          JSON.stringify(getInputValues())
+      )
+    } catch (e) {
+      setFormErrorMessage('Something bad happened: ' + e.toString())
+    }
+  }
+
+  return (
+    <form onSubmit={onSubmit}>
+      {children(
+        inputNamePropsTransformer,
```

Comment:
This is where I'm thinking we can hook into react-styleguidist and gets some nice docs explanation going here. I think it'll take java comment arbitrarily (maybe?). Alternatively, line 17 could have these instead. To restate, I think understanding these callback params will be really important for users.

Reply:
like we said on a call i think i'll just modify form-ii and we can work off that